### PR TITLE
lcm research

### PR DESCRIPTION
**Pull request by nebril**
This change is [Reviewable](https://reviewable.io/reviews/mirantis/k8s-appcontroller/101)

### CODE REVIEW

**Comment by jellonek** [comment]
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion.

---

*[docs/research/lcm.md, line 14 at r1](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101#-KW-RKoMLr6lLdFcBOZc:-KW-RKoMLr6lLdFcBOZd:bme6knq) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/0c67687f098ed81dfad34a3da453d6214c780d19/docs/research/lcm.md#L14)):*
> ```Markdown
> {
>   "onFail": [
>     {"type": "exec", "cmd": "promote_to_master.sh", "oneOf": ["slave1, slave2, slave3"}, #promote one of slaves to master
> ```

Small nit - please add space after # sign ;)
Also what are the entries in "oneOf"? Names of what?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101)*

**Comment by pigmej** [comment]
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion.

---

*[docs/research/lcm.md, line 14 at r1](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101#-KW-RKoMLr6lLdFcBOZc:-KW0ahrJmOlo63dvVFfq:buwiv1g) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/0c67687f098ed81dfad34a3da453d6214c780d19/docs/research/lcm.md#L14)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary>
> Small nit - please add space after # sign ;)
> Also what are the entries in "oneOf"? Names of what?
</details>
Those are names of vertices in graph I suppose, but yeah clarification is needed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101)*

**Comment by nebril** [comment]
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion.

---

*[docs/research/lcm.md, line 14 at r1](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101#-KW-RKoMLr6lLdFcBOZc:-KW23thN_NPjMn58co0m:b-896fix) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/0c67687f098ed81dfad34a3da453d6214c780d19/docs/research/lcm.md#L14)):*
<details><summary><i>Previously, pigmej (Jędrzej Nowak) wrote…</i></summary>
> Those are names of vertices in graph I suppose, but yeah clarification is needed.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101)*

**Comment by jellonek** [comment]
Reviewed 1 of 1 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[docs/research/lcm.md, line 14 at r2](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101#-KW286WiqImmofwkZd71:-KW286WiqImmofwkZd72:b-uny72t) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/8226439b451367eb4841efd005e3c2fb608e55b9/docs/research/lcm.md#L14)):*
> ```Markdown
> {
>   "onFail": [
>     {"type": "exec", "cmd": "promote_to_master.sh", "oneOf": ["slave1, slave2, slave3"}, # promote one of slaves to master
> ```

Missing closing `]` in list of slaves. When new slave is created - it's appended to this list, right?

---

*[docs/research/lcm.md, line 15 at r2](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101#-KW28Zh7bvJMkUnnwL0T:-KW28Zh7bvJMkUnnwL0U:b-mhqrvi) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/8226439b451367eb4841efd005e3c2fb608e55b9/docs/research/lcm.md#L15)):*
> ```Markdown
>   "onFail": [
>     {"type": "exec", "cmd": "promote_to_master.sh", "oneOf": ["slave1, slave2, slave3"}, # promote one of slaves to master
>     {"type": "create", "template": "slave pod template"} # create new slave pod which will replicate new master
> ```

What is this template? Template for pod name (like "slave%d")? Or template for whole pod definition, including somewhere inside template for name.
How this template is used, what variables will be expanded with which values? Maybe little example?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101)*

**Comment by nebril** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[docs/research/lcm.md, line 14 at r2](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101#-KW286WiqImmofwkZd71:-KW2CjO2Ph37Slc35XUG:b-896fix) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/8226439b451367eb4841efd005e3c2fb608e55b9/docs/research/lcm.md#L14)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary>
> Missing closing `]` in list of slaves. When new slave is created - it's appended to this list, right?
</details>
Done.

---

*[docs/research/lcm.md, line 15 at r2](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101#-KW28Zh7bvJMkUnnwL0T:-KW2CkmANwkQRvzfjdDJ:b-896fix) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/8226439b451367eb4841efd005e3c2fb608e55b9/docs/research/lcm.md#L15)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary>
> What is this template? Template for pod name (like "slave%d")? Or template for whole pod definition, including somewhere inside template for name.
> How this template is used, what variables will be expanded with which values? Maybe little example?
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101)*

**Comment by jellonek** [comment]
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101)*

File: docs/research/lcm.md [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+AC will need additional process traversing the deployment graph and checking the state of vertices. If the vertex is not ready and has deployed children, we need to take action, which will be defined in an annotation in the pod. We need a DSL for defining these actions, which will allow AC to read specification of these action and run them. This could be the implementation of `failurePolicy` from failure-handling research.
+
+Suggested annotation format:
+```
+{
+  "onFail": [
+    {"type": "exec", "cmd": "promote_to_master.sh", "oneOf": ["slave1, slave2, slave3"]}, # promote one of slaves to master
+    {"type": "create", "template": "slave pod template"} # create new slave pod which will replicate new master
+  ]
```
Comment (dshulyak) [inline docs/research/lcm.md]: frankly speaking it feels weird that you want to embed master election into appcontroller

Comment (dshulyak) [inline docs/research/lcm.md]: for instance how you will guarantee that previous master will become slave after network partition will be fixed, from above description i think you will end up with two masters.. 

i don't know much about mysql multislave, but if doesn't have any master election algorithm in it - it needs to be added on top with etcd/pacemaker

Comment (nebril) [inline docs/research/lcm.md]: TBH I did not take into account network partition into account. This whole mechanism is made with master pod crashing and being removed by k8s in mind. @pigmej should we consider network partition for this case?

Comment (jellonek) [inline docs/research/lcm.md]: As this is only an example usage and not full featured solution - imo netsplit is out of scope there, but yeah, we should wait for @pigmej.

**Comment by gluke77** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[docs/research/lcm.md, line 14 at r3](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101#-KW36zJOqssVbWiqU9tm:-KW36zJPwj7_vszoGecE:b-zbmv7s) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/9d18d105a182a921a7094c4bd244d3ee681f0441/docs/research/lcm.md#L14)):*
> ```Markdown
> {
>   "onFail": [
>     {"type": "exec", "cmd": "promote_to_master.sh", "oneOf": ["slave1, slave2, slave3"]}, # promote one of slaves to master
> ```

My concern here is that it looks like a limited time solution: if "oneOf" only contains the list of initially available slaves and newly created ones do not get into it, we'll eventually run out of slaves for the next master promotion.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101)*

**Comment by nebril** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[docs/research/lcm.md, line 14 at r3](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101#-KW36zJOqssVbWiqU9tm:-KW39LLrCoiGX8ifWhBm:b-uin80z) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/9d18d105a182a921a7094c4bd244d3ee681f0441/docs/research/lcm.md#L14)):*
<details><summary><i>Previously, gluke77 (Illia Khudoshyn) wrote…</i></summary>
> My concern here is that it looks like a limited time solution: if "oneOf" only contains the list of initially available slaves and newly created ones do not get into it, we'll eventually run out of slaves for the next master promotion.
</details>
That's true. I think the remedy for that would be to use k8s selectors. I will update the docs. Thanks for the feedback!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101)*

**Comment by nebril** [comment]
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions.

---

*[docs/research/lcm.md, line 14 at r1](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101#-KW-RKoMLr6lLdFcBOZc:-KW3AKpCJKANRpR_oXxP:b-896fix) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/0c67687f098ed81dfad34a3da453d6214c780d19/docs/research/lcm.md#L14)):*
<details><summary><i>Previously, nebril (Maciej Kwiek) wrote…</i></summary>
> Done.
</details>
Done.

---

*[docs/research/lcm.md, line 14 at r3](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101#-KW36zJOqssVbWiqU9tm:-KW3AKWC7a2ifdiuvz6w:b-896fix) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/9d18d105a182a921a7094c4bd244d3ee681f0441/docs/research/lcm.md#L14)):*
<details><summary><i>Previously, nebril (Maciej Kwiek) wrote…</i></summary>
> That's true. I think the remedy for that would be to use k8s selectors. I will update the docs. Thanks for the feedback!
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101)*

File: docs/research/lcm.md (hunk 1 above)
Comment (pigmej) [inline docs/research/lcm.md]: network partition problem is must have there, we need to solve it there. The solution *shouldnt* allow for multiple masters.

On master failure we need to:
- select new master
- be sure that when "current" master is back online it will be discarded.
And no we cannot rely on k8s mechanisms for that because of exactly the problem that @dshulyak mentioned

What we're also missing: 
- we need to be able to connect to mysql, therefore we need at least 2 k8s services, one for master one for slaves, it needs to be updated too, it's not described there at all.

Comment (pigmej) [inline docs/research/lcm.md]: @dshulyak answering your question, nothing very bad will happen, because user will be redirected to "new" master (user should connect to mysql_matser service), and even if old master will be online, noone will be able to connect to it (obviously not immediately), and slaves will talk to new master, so nothing wrong in long term.

**Comment by pigmej** [comment]
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions.

---

*[docs/research/lcm.md, line 14 at r1](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101#-KW-RKoMLr6lLdFcBOZc:-KW3eUo1CLZZ0Lm6Dcx6:b2a3zez) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/0c67687f098ed81dfad34a3da453d6214c780d19/docs/research/lcm.md#L14)):*
<details><summary><i>Previously, nebril (Maciej Kwiek) wrote…</i></summary>
> Done.
</details>
what is `role:slave` ?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101)*

**Comment by dshulyak** [comment]
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions.

---

*[docs/research/lcm.md, line 16 at r3](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101#-KVyCfAu17Ej2cvQKfZM-r3-16:-KW3k3GvavZZtpjKynid:b-mg8vc) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/9d18d105a182a921a7094c4bd244d3ee681f0441/docs/research/lcm.md#L16)):*
<details><summary><i>Previously, pigmej (Jędrzej Nowak) wrote…</i></summary>
> @dshulyak answering your question, nothing very bad will happen, because user will be redirected to "new" master (user should connect to mysql_matser service), and even if old master will be online, noone will be able to connect to it (obviously not immediately), and slaves will talk to new master, so nothing wrong in long term.
</details>
btw what are the reasons to embed this into appcontroller?

i would add this piece of logic as a separate pod monitor/controller

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101)*

**Comment by zen** [comment]
*[docs/research/lcm.md, line 16 at r3](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101#-KVyCfAu17Ej2cvQKfZM-r3-16:-KW4aIPYD2K9MQj8n9Qs:bg69jln) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/9d18d105a182a921a7094c4bd244d3ee681f0441/docs/research/lcm.md#L16)):*
<details><summary><i>Previously, dshulyak (Dmitry Shulyak) wrote…</i></summary>
> btw what are the reasons to embed this into appcontroller?
> 
> i would add this piece of logic as a separate pod monitor/controller
</details>
Proper fencing or STONITH is must have,  otherwise you can expect all sorts of strange issues.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101)*

**Comment by pigmej** [comment]
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions.

---

*[docs/research/lcm.md, line 16 at r3](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101#-KVyCfAu17Ej2cvQKfZM-r3-16:-KW4oSsTqgr6pVbfS9lN:b-5t9ccw) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/9d18d105a182a921a7094c4bd244d3ee681f0441/docs/research/lcm.md#L16)):*
<details><summary><i>Previously, zen (Tomasz Napierala) wrote…</i></summary>
> Proper fencing or STONITH is must have,  otherwise you can expect all sorts of strange issues.
</details>
@zen 
Yes but you will get it by the nature of k8s services etc. 
Basically in this mysql case you need 2 (or 3) services, one for master one for slaves (or separate for primary and secondary slave), slaves are configured to replicate from master service (not master pod), therefore afetr you switch service you have guarantee that all will replicate from that master. Also after you switch master service user will connect to proper master server without possibility to connect to old one. Obviously there is still time required to properly spread that information through k8s cluster but you can also take some other steps to minimize it almost completely (we can spin off new pod for slave, and request removal of existing master and remove current master from all k8s services etc)

@nebril and that's what I'm missing in that research. Executing these bash scripts is not enough. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101)*

**Comment by pigmej** [comment]
@nebril moreover I'm pretty sure that approach with "scripts" is too limited/too hacky. For example how I'm supposed to do:
- propagate primary slave to new master, it requires changes in this primary slave configuration
- how can I see which of them is primary slave which is secondary? In the demo that you provided you do some `oneOf` which I think is wrong in this case
- how am I supposed to express "wipe data of that master and make it secondary slave" (which is pretty much the same as create new pod, forgot about previous one)

I really think that we need subgraph / graph execution there.

**Comment by nebril** [comment]
I agree that we should be able to execute graphs/subgraphs. We can put it as one of actions to do on failure. I updated the example spec to show possible definition of subgraph rerun step.

---

Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions.

---

*[docs/research/lcm.md, line 16 at r3](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101#-KVyCfAu17Ej2cvQKfZM-r3-16:-KW7rNvE9Q1N_YzgdQtn:b291p2g) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/9d18d105a182a921a7094c4bd244d3ee681f0441/docs/research/lcm.md#L16)):*
<details><summary><i>Previously, pigmej (Jędrzej Nowak) wrote…</i></summary>
> @zen 
> Yes but you will get it by the nature of k8s services etc. 
> Basically in this mysql case you need 2 (or 3) services, one for master one for slaves (or separate for primary and secondary slave), slaves are configured to replicate from master service (not master pod), therefore afetr you switch service you have guarantee that all will replicate from that master. Also after you switch master service user will connect to proper master server without possibility to connect to old one. Obviously there is still time required to properly spread that information through k8s cluster but you can also take some other steps to minimize it almost completely (we can spin off new pod for slave, and request removal of existing master and remove current master from all k8s services etc)
> 
> @nebril and that's what I'm missing in that research. Executing these bash scripts is not enough. 
</details>
@dshulyak the reason to put this into ac is that ac should be responsible for error handling within the deployment graph. Handling failures should be part of graph definition, which is responsibility of ac. I think that we should go with whole error handling story first, keep in mind the lcm and then see what we need to add to error handling to actually handle lcm.

@jnowak about deleting the master - we can run a job which will finish by deleting master and make it part of deployment graph.
About two services - why do we need a service for slaves? Slaves will replicate master via master service, and I agree that we should have this one, but I don't see the benefit of service in front of slave pods.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101#-:-KW7qmkjX657JNVrz5ew:bgrwhv6)*

**Comment by nebril** [comment]
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions.

---

*[docs/research/lcm.md, line 14 at r1](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101#-KW-RKoMLr6lLdFcBOZc:-KW88OvSst7vi7t9B6y9:b6her6a) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/0c67687f098ed81dfad34a3da453d6214c780d19/docs/research/lcm.md#L14)):*
<details><summary><i>Previously, pigmej (Jędrzej Nowak) wrote…</i></summary>
> what is `role:slave` ?
</details>
It's a k8s selector.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101)*

**Comment by pigmej** [comment]
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions.

---

*[docs/research/lcm.md, line 14 at r1](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101#-KW-RKoMLr6lLdFcBOZc:-KW9FwaGgtDKQRdXLn6_:bb7p5f3) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/0c67687f098ed81dfad34a3da453d6214c780d19/docs/research/lcm.md#L14)):*
<details><summary><i>Previously, nebril (Maciej Kwiek) wrote…</i></summary>
> It's a k8s selector.
</details>
thx.

---

*[docs/research/lcm.md, line 16 at r3](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101#-KVyCfAu17Ej2cvQKfZM-r3-16:-KW9Fzmn-zLYtRM13wPN:bhdz6fk) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/9d18d105a182a921a7094c4bd244d3ee681f0441/docs/research/lcm.md#L16)):*
<details><summary><i>Previously, nebril (Maciej Kwiek) wrote…</i></summary>
> @dshulyak the reason to put this into ac is that ac should be responsible for error handling within the deployment graph. Handling failures should be part of graph definition, which is responsibility of ac. I think that we should go with whole error handling story first, keep in mind the lcm and then see what we need to add to error handling to actually handle lcm.
> 
> @jnowak about deleting the master - we can run a job which will finish by deleting master and make it part of deployment graph.
> About two services - why do we need a service for slaves? Slaves will replicate master via master service, and I agree that we should have this one, but I don't see the benefit of service in front of slave pods.
</details>
you need service for slave to allow user to connect into it :) 

And please remember that there are 2 types of "slave", primary slave and secondary slave, secondary slaves are replicating from primary slave, they need to know where is "new primary slave". This is how it should be designed. You have one master, one primary slave, and multiple secondary slaves, master => p.slave => s.slaves, not master => slaves.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101)*

**Comment by nebril** [comment]
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions.

---

*[docs/research/lcm.md, line 16 at r3](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101#-KVyCfAu17Ej2cvQKfZM-r3-16:-KWCuqThiHaHd3x2GObU:b-q7r3i5) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/9d18d105a182a921a7094c4bd244d3ee681f0441/docs/research/lcm.md#L16)):*
<details><summary><i>Previously, pigmej (Jędrzej Nowak) wrote…</i></summary>
> you need service for slave to allow user to connect into it :) 
> 
> And please remember that there are 2 types of "slave", primary slave and secondary slave, secondary slaves are replicating from primary slave, they need to know where is "new primary slave". This is how it should be designed. You have one master, one primary slave, and multiple secondary slaves, master => p.slave => s.slaves, not master => slaves.
</details>
I think that if users are supposed to read from slaves, there should be a load balancer service which will select both master and slaves. So indeed we whould have two services. One for writing - which will only select master and one for reading - which will select all (or only slaves if we don't want to read from master).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101)*

**Comment by pigmej** [comment]
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions.

---

*[docs/research/lcm.md, line 16 at r3](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101#-KVyCfAu17Ej2cvQKfZM-r3-16:-KWESqFgb_xKL8CC3GWU:bcop8vb) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/9d18d105a182a921a7094c4bd244d3ee681f0441/docs/research/lcm.md#L16)):*
<details><summary><i>Previously, nebril (Maciej Kwiek) wrote…</i></summary>
> I think that if users are supposed to read from slaves, there should be a load balancer service which will select both master and slaves. So indeed we whould have two services. One for writing - which will only select master and one for reading - which will select all (or only slaves if we don't want to read from master).
</details>
And that's more or less the idea. The whole idea is to allow users to read from slaves ;-) And yeah it should be k8s service. We shouldn't *allow* to read from both master and slave at the same time (user may want it though). Still we will probably need to know what servers are where so we may need more stuff, but it's separate story and it doesn't complicate stuff at all.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101)*

**Comment by nebril** [comment]
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions.

---

*[docs/research/lcm.md, line 14 at r1](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101#-KW-RKoMLr6lLdFcBOZc:-KWhEU9Sc54IZGWuJuiz:b-896fix) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/0c67687f098ed81dfad34a3da453d6214c780d19/docs/research/lcm.md#L14)):*
<details><summary><i>Previously, pigmej (Jędrzej Nowak) wrote…</i></summary>
> thx.
</details>
Done.

---

*[docs/research/lcm.md, line 16 at r3](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101#-KVyCfAu17Ej2cvQKfZM-r3-16:-KWhE_eZ6GLOx-ypPq1X:b-896fix) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/9d18d105a182a921a7094c4bd244d3ee681f0441/docs/research/lcm.md#L16)):*
<details><summary><i>Previously, pigmej (Jędrzej Nowak) wrote…</i></summary>
> And that's more or less the idea. The whole idea is to allow users to read from slaves ;-) And yeah it should be k8s service. We shouldn't *allow* to read from both master and slave at the same time (user may want it though). Still we will probably need to know what servers are where so we may need more stuff, but it's separate story and it doesn't complicate stuff at all.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101)*

**Comment by pigmej** [comment]
Review status: 0 of 1 files reviewed at latest revision, 4 unresolved discussions.

---

*[docs/research/lcm.md, line 16 at r4](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101#-KWhSz_6wAXe6O9VRbyd:-KWhSz_7T3A0W-_uL2zu:bl7ns8j) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/bb314e79ec9967e1774e1f6c45b16038cea21e76/docs/research/lcm.md#L16)):*
> ```Markdown
>     {"type": "exec", "cmd": "promote_to_master.sh", "oneOf": "role:slave"}, # promote one of slaves to master
>     {"type": "create", "template": "slave pod template"}, # create new slave pod which will replicate new master
> 	{"type": "rerun", "root": "subgraph-vertice-id"} # redeploy subgraph
> ```

tab vs spaces? (https://www.youtube.com/watch?v=SsoOG6ZeyUI)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101)*

**Comment by nebril** [comment]
Review status: 0 of 1 files reviewed at latest revision, 4 unresolved discussions.

---

*[docs/research/lcm.md, line 16 at r4](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101#-KWhSz_6wAXe6O9VRbyd:-KWhVEIR2Kp7E20sxP-b:b-896fix) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/bb314e79ec9967e1774e1f6c45b16038cea21e76/docs/research/lcm.md#L16)):*
<details><summary><i>Previously, pigmej (Jędrzej Nowak) wrote…</i></summary>
> tab vs spaces? (https://www.youtube.com/watch?v=SsoOG6ZeyUI)
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101)*

**Comment by pigmej** [comment]
Review status: 0 of 1 files reviewed at latest revision, 5 unresolved discussions.

---

*[docs/research/lcm.md, line 3 at r5](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101#-KX5zh-11nmqVr_u_SFd:-KX5zh-11nmqVr_u_SFe:b-yas1kt) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/2bb0f351bee99b15f07c2e1bda63015705262538/docs/research/lcm.md#L3)):*
> one of slave nodes
In case of failure we want to promote primary slave to new master, and select new primary slave from secondary slaves. Also we need to ensure that old master will join as secondary slave after it will be back.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101)*

**Comment by pigmej** [comment]
Review status: 0 of 1 files reviewed at latest revision, 6 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101#-KXARtT57Td7RmQP4w8Y:-KXARtT6XIJydXwZd2h4:bk95g9w) (no related file):*
@nebril Sorry for late feedback but we're missing there some grace periods. We should have possibility to configure it.  

We should have at least (times and numbers are just examples, you know what I may mean there):
- 3 consecutive failures in 1 minute => execute error
- 3 consecutive failures in 10 seconds => execute error

We should also have possibility to specify:
- Ignore errors for 10 seconds after start
- Ignore errors for 1 minute after start


So basically we should count errors and collect them, instead of executing hard action on first spotted failure (may be temp network error)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101)*

**Comment by nebril** [comment]
Review status: 0 of 1 files reviewed at latest revision, 6 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101#-KXARtT57Td7RmQP4w8Y:-KXArEFookJH9vtiacbI:brv0yth) (no related file):*
<details><summary><i>Previously, pigmej (Jędrzej Nowak) wrote…</i></summary>
> @nebril Sorry for late feedback but we're missing there some grace periods. We should have possibility to configure it.  
> 
> We should have at least (times and numbers are just examples, you know what I may mean there):
> - 3 consecutive failures in 1 minute => execute error
> - 3 consecutive failures in 10 seconds => execute error
> 
> We should also have possibility to specify:
> - Ignore errors for 10 seconds after start
> - Ignore errors for 1 minute after start
> 
> 
> So basically we should count errors and collect them, instead of executing hard action on first spotted failure (may be temp network error)
</details>
I agree with and understand the first one, but how do you want to specify the "start" moment? Creation of pod? End of graph deployment?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101)*

**Comment by pigmej** [comment]
Review status: 0 of 1 files reviewed at latest revision, 6 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101#-KXARtT57Td7RmQP4w8Y:-KXAsL9j2NPvD3lxlUGI:b36f3f0) (no related file):*
<details><summary><i>Previously, nebril (Maciej Kwiek) wrote…</i></summary>
> I agree with and understand the first one, but how do you want to specify the "start" moment? Creation of pod? End of graph deployment?
</details>
First check == start, after you execute "hook" you reset check counter.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101)*

**Comment by nebril** [comment]
Review status: 0 of 1 files reviewed at latest revision, 6 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101#-KXARtT57Td7RmQP4w8Y:-KXAstPBzeZrSC5tZk03:b-7yu7nj) (no related file):*
<details><summary><i>Previously, pigmej (Jędrzej Nowak) wrote…</i></summary>
> First check == start, after you execute "hook" you reset check counter.
</details>
Wilco

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101)*

**Comment by nebril** [comment]
Review status: 0 of 1 files reviewed at latest revision, 6 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101#-KXARtT57Td7RmQP4w8Y:-KXB6eN1_icAxLTH8IQk:b-j331xl) (no related file):*
<details><summary><i>Previously, nebril (Maciej Kwiek) wrote…</i></summary>
> Wilco
</details>
On second thought: yeah, its place is not in this research
it should be a separate issue on GH, e.g.
“Make status checks take consecutive failures in time into account to get rid of temporary failures"

Will create the issue

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101)*

**Comment by gluke77** [comment]
:lgtm:

---

Review status: 0 of 1 files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101#-:-KXGAWHi6K4PcvnNMrtp:bnfp4nl)*

**Comment by pigmej** [comment]
Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101)*

**Comment by pigmej** [comment]
:lgtm: 

---

Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101#-:-KXGBo7bxAD8oGXkVF0K:bg7igkv)*

**Comment by pigmej** [comment]
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/101)*
